File: vulnerable_app.php
<?php

// Hardcoded credentials (CWE-798)
define('DB_PASS', 'php_hardcoded_pass_123');
define('SECRET_KEY', 'php_secret_key_abcdef1234567890');
define('API_TOKEN', 'Bearer eyJhbGciOiJub25lIn0.hardcoded.token');

// SQL Injection (CWE-89)
function getUser($username) {
    $conn = new mysqli("localhost", "root", DB_PASS, "mydb");
    $result = $conn->query("SELECT * FROM users WHERE username = '$username'");
    return $result->fetch_all();
}

function getProduct($id) {
    $conn = new mysqli("localhost", "root", DB_PASS, "mydb");
    $result = $conn->query("SELECT * FROM products WHERE id = " . $id);
    return $result->fetch_all();
}

function deleteUser($id) {
    $conn = new mysqli("localhost", "root", DB_PASS, "mydb");
    $conn->query("DELETE FROM users WHERE id = " . $_GET['id']);
}

// Command Injection (CWE-78)
function pingHost($host) {
    echo shell_exec("ping -c 1 " . $host);
}

function runCommand($cmd) {
    return system($_GET['cmd']);
}

function listFiles($dir) {
    return exec("ls " . $dir);
}

// XSS (CWE-79)
function greetUser($name) {
    echo "<h1>Hello " . $name . "</h1>";
}

function showSearch($query) {
    echo "<p>Results for: " . $_GET['q'] . "</p>";
}

// Path Traversal (CWE-22)
function readFile($filename) {
    return file_get_contents("/var/app/" . $filename);
}

function includeFile($page) {
    include("/pages/" . $_GET['page'] . ".php");
}

// Weak cryptography (CWE-326)
function hashPassword($password) {
    return md5($password);
}

function weakEncrypt($data) {
    return sha1($data);
}

// Insecure deserialization (CWE-502)
function loadObject($data) {
    return unserialize($data);
}

function loadUserPrefs() {
    return unserialize($_COOKIE['prefs']);
}

// Insecure random (CWE-330)
function generateToken() {
    return rand(0, 999999);
}

function generateSessionId() {
    return mt_rand();
}

// SSRF (CWE-918)
function fetchURL($url) {
    return file_get_contents($_GET['url']);
}

// File upload without validation (CWE-434)
function uploadFile() {
    move_uploaded_file($_FILES['file']['tmp_name'], '/uploads/' . $_FILES['file']['name']);
}

// Open redirect (CWE-601)
function redirect() {
    header("Location: " . $_GET['url']);
}

// Code injection via eval (CWE-95)
function calculate($expr) {
    eval('$result = ' . $expr . ';');
    return $result;
}

// Sensitive data in logs (CWE-532)
function login($user, $password) {
    error_log("Login: user=$user password=$password");
    return $password == DB_PASS;
}

// XXE via simplexml (CWE-611)
function parseXML($xmlString) {
    return simplexml_load_string($xmlString, 'SimpleXMLElement', LIBXML_NOENT);
}

// LDAP Injection (CWE-90)
function ldapSearch($username) {
    $ldap = ldap_connect("ldap://localhost");
    ldap_search($ldap, "dc=example,dc=com", "uid=" . $username);
}

// Sensitive data in URL (CWE-598)
function sendRequest($token) {
    file_get_contents("https://api.example.com/data?token=" . $token . "&user=" . DB_PASS);
}

?>

File: vulnerable_app.pl
#!/usr/bin/perl
use strict;
use warnings;
use DBI;
use CGI;

# Hardcoded credentials (CWE-798)
my $DB_PASS   = "perl_hardcoded_pass_123";
my $API_KEY   = "perl_api_key_hardcoded_abcdef";
my $SECRET    = "perl_secret_key_never_rotate";

my $cgi = CGI->new;

# SQL Injection (CWE-89)
sub get_user {
    my ($username) = @_;
    my $dbh = DBI->connect("dbi:mysql:mydb", "root", $DB_PASS);
    my $query = "SELECT * FROM users WHERE username = '$username'";
    return $dbh->selectall_arrayref($query);
}

sub search_records {
    my ($term) = @_;
    my $dbh = DBI->connect("dbi:mysql:mydb", "root", $DB_PASS);
    $dbh->do("SELECT * FROM records WHERE name LIKE '%$term%'");
}

# Command Injection (CWE-78)
sub ping_host {
    my ($host) = @_;
    system("ping -c 1 $host");              # Unsanitized
}

sub run_command {
    my ($cmd) = @_;
    return `$cmd`;                          # Backtick execution
}

sub list_files {
    my ($dir) = @_;
    return qx{ls $dir};                    # Shell injection
}

# Path Traversal (CWE-22)
sub read_file {
    my ($filename) = @_;
    open(my $fh, "<", "/var/app/$filename") or die;
    return do { local $/; <$fh> };
}

# XSS (CWE-79)
sub greet_user {
    my ($name) = $cgi->param("name");
    print "<h1>Hello $name</h1>";           # No encoding
}

sub show_results {
    my $q = $cgi->param("q");
    print "<p>Results for: $q</p>";
}

# Insecure eval (CWE-95)
sub calculate {
    my ($expr) = @_;
    return eval($expr);                     # Arbitrary code execution
}

# Weak cryptography - use of crypt() with DES (CWE-326)
sub hash_password {
    my ($password) = @_;
    return crypt($password, "salt");        # DES-based, weak
}

# Sensitive data in logs (CWE-532)
sub login {
    my ($user, $pass) = @_;
    warn "Login: user=$user pass=$pass\n";  # Credentials in logs
    return $pass eq $DB_PASS;
}

# Open redirect (CWE-601)
sub redirect_user {
    my $url = $cgi->param("url");
    print $cgi->redirect($url);             # Unvalidated redirect
}

# Insecure temp file (CWE-377)
sub create_temp {
    my $tmpfile = "/tmp/app_$$";            # Predictable temp file
    open(my $fh, ">", $tmpfile);
    return $fh;
}

# File inclusion (CWE-98)
sub include_module {
    my ($module) = $cgi->param("module");
    require $module;                        # Remote file inclusion risk
}

# LDAP Injection (CWE-90)
sub ldap_search {
    my ($username) = @_;
    my $filter = "uid=$username";           # Unsanitized LDAP filter
    # $ldap->search(filter => $filter);
    return $filter;
}

# Main execution
my $username = $cgi->param("username") || "admin' OR '1'='1";
get_user($username);
ping_host($cgi->param("host") || "localhost");
calculate($cgi->param("expr") || "1+1");
